rust_src/src/process.rs

```diff
@@ -102,6 +116,102 @@ pub fn set_process_plist(process: LispObject, plist: LispObject) -> LispObject {
     }
 }
 
+/// Return the status of PROCESS.
+/// The returned value is one of the following symbols:
+/// run  -- for a process that is running.
+/// stop -- for a process stopped but continuable.
+/// exit -- for a process that has exited.
+/// signal -- for a process that has got a fatal signal.
+/// open -- for a network stream connection that is open.
+/// listen -- for a network stream server that is listening.
+/// closed -- for a network stream connection that is closed.
+/// connect -- when waiting for a non-blocking connection to complete.
+/// failed -- when a non-blocking connection has failed.
+/// nil -- if arg is a process name and no such process exists.
+/// PROCESS may be a process, a buffer, the name of a process, or
+/// nil, indicating the current buffer's process.
+#[lisp_fn]
+pub fn process_status(process: LispObject) -> LispObject {
+    let mut p = process;
+    if p.is_string() {
+        p = get_process(p);
+    } else {
+        p = unsafe { LispObject::from(cget_process(p.to_raw())) };
+    }
+    if p.is_nil() {
+        return p;
+    }
+    let p_ref = p.as_process_or_error();
+    if unsafe { pget_raw_status_new(p_ref.as_ptr()) } != 0 {
+        unsafe { update_status(p_ref.as_ptr()) };
+    }
+    let mut status = LispObject::from(p_ref.status);
+    if let Some(c) = status.as_cons() {
+        status = LispObject::from(c.car());
+    };
+    let process_type = LispObject::from(p_ref.process_type);
+    if process_type.eq(LispObject::from(Qnetwork)) || process_type.eq(LispObject::from(Qserial))
+        || process_type.eq(LispObject::from(Qpipe))
+    {
+        let process_command = LispObject::from(p_ref.command);
+        if status.eq(LispObject::from(Qexit)) {
+            status = LispObject::from(Qclose);
+        } else if process_command.eq(LispObject::constant_t()) {
+            status = LispObject::from(Qstop);
+        } else if status.eq(LispObject::from(Qrun)) {
+            status = LispObject::from(Qopen);
+        }
+    }
+    status
+}
+
+/// Set buffer associated with PROCESS to BUFFER (a buffer, or nil).
+/// Return BUFFER.
+#[lisp_fn]
+fn set_process_buffer(process: LispObject, buffer: LispObject) -> LispObject {
+    let mut p_ref = process.as_process_or_error();
+    if buffer.is_not_nil() {
+        buffer.as_buffer_or_error();
+    }
+    p_ref.set_buffer(buffer);
+    let process_type = LispObject::from(p_ref.process_type);
+    if process_type.eq(LispObject::from(Qnetwork)) || process_type.eq(LispObject::from(Qserial))
```

Comment:
Why drop the named predicates?

Reply:
You mean the C macros `SERIALCONN1_P` etc? It's used in 5 places, so it's a tradeoff between being explicit and being concise. I'm open to persuasion either way :)

Reply:
Would it be a problem to compare `Lisp_Object`s here to avoid the casting ?

Reply:
My comment went out of band.

Keeping the name should make rollups easier. I am predominantly a Python code. Explicit is always better than implicit. Names have value

Reply:
I think `Qserial` etc. is explicit enough for now, but the if statement is too verbose for what is done here.